tests/integ/datasource/test_postgres.py

```diff
@@ -0,0 +1,428 @@
+#
+# Copyright (c) 2012-2025 Snowflake Computing Inc. All rights reserved.
+#
+
+import pytest
+
+import psycopg2  # noqa: F401
+import pandas  # noqa: F401
+
+from snowflake.snowpark import Row
+from snowflake.snowpark._internal.data_source.drivers import Psycopg2Driver
+from snowflake.snowpark._internal.data_source.drivers.psycopg2_driver import (
+    Psycopg2TypeCode,
+)
+from snowflake.snowpark._internal.data_source.utils import DBMS_TYPE
+from snowflake.snowpark.exceptions import SnowparkDataframeReaderException
+from snowflake.snowpark.types import (
+    DecimalType,
+    BinaryType,
+    VariantType,
+    StructType,
+    StructField,
+    StringType,
+    TimeType,
+    BooleanType,
+    IntegerType,
+    FloatType,
+    DoubleType,
+    DateType,
+    TimestampType,
+    TimestampTimeZone,
+)
+from snowflake.snowpark._internal.data_source.dbms_dialects.postgresql_dialect import (
+    PostgresDialect,
+)
+from tests.parameters import POSTGRES_CONNECTION_PARAMETERS
+from tests.resources.test_data_source_dir.test_postgres_data import (
+    POSTGRES_TABLE_NAME,
+    EXPECTED_TEST_DATA,
+    EXPECTED_TYPE,
+    POSTGRES_TEST_EXTERNAL_ACCESS_INTEGRATION,
+)
+from tests.utils import IS_IN_STORED_PROC
+
+DEPENDENCIES_PACKAGE_UNAVAILABLE = False
+
+
+pytestmark = [
+    pytest.mark.skipif(DEPENDENCIES_PACKAGE_UNAVAILABLE, reason="Missing 'psycopg2'"),
+    pytest.mark.skipif(IS_IN_STORED_PROC, reason="Need External Access Integration"),
+]
+
+
+def create_postgres_connection():
+    return psycopg2.connect(**POSTGRES_CONNECTION_PARAMETERS)
+
+
+@pytest.mark.parametrize(
+    "input_type, input_value",
+    [
+        ("table", POSTGRES_TABLE_NAME),
+        ("query", f"(SELECT * FROM {POSTGRES_TABLE_NAME})"),
```

Comment:
is it expected that user would input `(select * from TABLE)` instead of `select * from TABLE` ? I know that pyspark support input of `select * from TABLE`

Reply:
yes, this is expected
postgresql does not support `select * from select * from table`